### PR TITLE
Fixed non-determinism in 21-capture-in-cobegin et al.

### PR DESCRIPTION
**Pull request by vasslitvinov**
JIRA Issue 63: https://chapel.atlassian.net/browse/CHAPEL-63

[To the reviewer: I will remove the #include that I left in there unintentionally.]

THE ISSUE

In the following code

```
    var x = 5;
    ....
    coforall .... do
      writeln(x);
```

we want all iterations of coforall to print the value of 'x'
that 'x' had immediately before entering 'coforall'.
As a result, all iterations are guaranteed to print the same.

Prior to this change, our implementation was capturing
the value of 'x' at the time each coforall task was created,
i.e. when initializing each argument bundle.
So it was possible that 'x' changes between initializations
of the first and second bundle, violating our desired property.

Here are a couple of fleshed-out examples:

```
    var x = 5;
    begin with (ref x) { x = 6; }
    coforall ... do writeln(x);
```

```
    var x = 5;
    proc update() { x = 6; }
    cobegin { /*Task1*/writeln(x); /*Task2*/update(); }
```

THE FIX

For each variable to be captured, I create a temp and initialize
it from that variable right before the start of the corresponding
task construct. That temp is used instead of the original variable
when initializing the argument bundles.

WHAT'S UP WITH RECORDS?

Right now the blank task intent for records is incorrectly implemented
as 'const in'. Since this is incorrect, I explicitly exclude records
from The Fix above. Therefore the records continue to be captured at
task creation time. I adjusted the tests not to test records, because
the behavior remains nondeterministic.

WHAT'S WITH STRINGS AND UNIONS?

What I wrote about records above should apply to strings and unions.
I do not know for sure whether it does. For now I exclude strings and
unions from testing as well, because most likely their behavior in
this regard currently is also most likely nondeterministic.

### CODE REVIEW

File: compiler/resolution/functionResolution.cpp [hunk 1]
```diff
@@ -2739,7 +2741,112 @@ getVisibleFunctions(BlockStmt* block,
...
+
+// block id -> (sym -> sym)
+typedef std::map<int, SymbolMap*> CapturedValueMap;
+static CapturedValueMap capturedValues;
+static void freeCache(CapturedValueMap& c) {
+  for (std::map<int, SymbolMap*>::iterator it = c.begin(); it != c.end(); ++it)
+    delete it->second;
+}
+
+static void captureTaskIntentValues(int argNum, ArgSymbol* formal,
```
Comment (hildeth) [inline compiler/resolution/functionResolution.cpp]: A header comment here, explaining the reason for the existence of this function, what it does, and why it is not called for all types would be *very* helpful....

**Comment by hildeth** [comment]
I just happened to take a quick look through this.  Aside from a missing comment regarding the semantics being implemented and the approach taken, this patch looks good to me.

I'll leave final approval up to @Kyle-B

**Comment by Kyle-B** [comment]
I think with the updates mentioned by Tom and yourself, (Comment and remove the header) things are fine.

**Comment by vasslitvinov** [comment]
Thanks @hildeth @Kyle-B - will do!

**Pull request merged**